### PR TITLE
Add domain validation in DomainSpecificString

### LINKED ISSUE #4088: createRoom fails with a 500 error if a badly-formatted mxid is in the `invitees` list
For example, if you pass an MXID which has a trailing space.

See https://github.com/matrix-org/synapse/issues/4032#issuecomment-429282870 for a stacktrace.

### PR DESCRIPTION
**Pull request by interru**
FIXES #4088

This adds an additional validation check whether the domain/hostname meets the criteria of a valid hostname

### CODE REVIEW

File: synapse/types.py [hunk 1]
```diff
@@ -138,24 +139,43 @@ def __deepcopy__(self, memo):
...
+                400, "Expected %s string to start with %r" % (
+                    cls.__name__, cls.SIGIL,
+                )
+            )
+
+        port = None
+        try:
+            localpart, domain = s[1:].split(':', 1)
+            if ':' in domain:
+                domain, port = domain.split(':')
```
Comment (richvdh) [inline synapse/types.py]: this looks like it will fail on IPv6 addresses.

**Comment by interru** [comment]
> The code you are changing here is on a hot path - we'll end up spending a non-trivial amount of resources validating server names which are already known to be valid. I'd prefer it if we validated identifiers at the point of entry to the system - for example, #4088 could be fixed with better checks in the RoomCreationHandler.

You have two options:
- Either we only accept ASCII/IDNA encoded URLs

**Comment by interru** [comment]
Wron button ;)

File: synapse/types.py (hunk 1 above)
Comment (interru) [reply synapse/types.py]: Good point! Completely forgot IPv6.

**Comment by interru** [comment]
> The code you are changing here is on a hot path - we'll end up spending a non-trivial amount of resources validating server names which are already known to be valid.

Thats raises the question why UserIDs (& co) get converted back to an normal string and parsed a second time?! The only places where we should get it as string and not as an instance is on IO. The only IO you can trust is the Database (assuming the data in the database was already validated).

In my opinion we also should only handle the domain in one form (IDNA or unicode). Currently both forms are accepted which could perhaps cause weird issues if someone mixes them.

I need to make myself more familiar with the code, spec and ecosystem. Could also be that I currently might seeing problems because of Dunning-Kruger effect. ;)

**Comment by richvdh** [comment]
> Thats raises the question why UserIDs (& co) get converted back to an normal string and parsed a second time?! The only places where we should get it as string and not as an instance is on IO. The only IO you can trust is the Database (assuming the data in the database was already validated).

well yes, it's specifically the path from the database that I was thinking of. However, I don't think it's as simple as you imagine: some parts of the code expect strings, and some expect objects, which naturally leads to a reasonable amount of conversion between the two. You could make this more consistent, but it's not obvious to me that doing so would reduce the amount of conversion to be done. For instance, there are a lot of simple CRUD paths which would end up converting string to instance and back to string again for no real value.

> In my opinion we also should only handle the domain in one form (IDNA or unicode). Currently both forms are accepted which could perhaps cause weird issues if someone mixes them.

https://matrix.org/docs/spec/appendices.html#server-name is pretty clear that they should be ascii (ie, IDNA.) Anywhere that accepts them as unicode is probably a bug.

**Comment by richvdh** [comment]
appears abandoned by contributor

**Pull request closed without merging**